### ISSUE TITLE
BUG: A segfault in chararray (2.0.0dev0 regression)

### ISSUE THREAD

**Issue by neutrinoceros** [numpy/numpy#25513]
### Describe the issue:

Astropy's tests against numpy's dev branch are currently choking on a segfault.
The issue can be boiled down to a pure numpy reproducer.

I wasn't able to setup my dev env properly to provide the full stack trace information yet (I'm working on it), but I could bisect the problem to https://github.com/numpy/numpy/commit/19396d298beca3f56d93d59e18b6965853c9eebc (https://github.com/numpy/numpy/pull/25171)

### Reproduce the code example:

```python
import numpy as np

carr = np.chararray((2,), itemsize=25)
carr[:] = [b'  4.52173913043478315E+00', b'  4.95652173913043548E+00']
carr.replace(b"E", b"D")
```

### Python and NumPy Versions:

2.0.0.dev0+git20231230.ee3124b
3.12.0 (main, Oct 27 2023, 11:50:57) [Clang 15.0.0 (clang-1500.0.40.1)]

### Runtime Environment:

```
[{'numpy_version': '2.0.0.dev0+git20231231.f1fac82',
  'python': '3.12.1 (main, Dec 11 2023, 18:41:50) [Clang 15.0.0 '
            '(clang-1500.0.40.1)]',
  'uname': uname_result(system='Darwin', node='kwanzaabot.home', release='23.2.0', version='Darwin Kernel Version 23.2.0: Wed Nov 15 21:55:06 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6020', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
None
```

### Context for the issue:

https://github.com/astropy/astropy/issues/15797

**Comment by mhvk** [comment]
Trying a bit, the segfault does not always happen immediately, but the results definitely are not always correct:
```
carr = np.chararray((2,), itemsize=25)
carr[:] = [b'  4.52173913043478315E+00', b'  4.95652173913043548E+00']
carr.replace(b'E', b'D')
# array([b'  4.52173913043478315E+00', b'DDDDDDDDDDDDDDDDDDDDDD+00'],
#       dtype='|S25')
```
When things don't give a segfault, on exit of ipython one gets "double free or corruption (out)".

**Comment by mhvk** [comment]
Found the culprit if not yet the solution - the special-cased single-character case, and specifically the case where `memchr` is invoked; see
https://github.com/numpy/numpy/blob/bab72802d9bfa3b17b2abbb0eade7894ff8c53bf/numpy/_core/src/umath/string_fastsearch.h#L59-L78

And indeed, `.find()` is broken too:
```
arr.find(b'E')
# array([0, 0])
```

Basically, `s.buf` gets overwritten to be the same as `p.buf`.

**Comment by mhvk** [comment]
Also, is it intended that if the replacement string is too long to possibly be present, that an empty output is returned?
```
In [1]: a = np.array('abc').view(np.char.chararray)

In [2]: a.replace('012345', '')
Out[2]: array('', dtype='<U3')

In [3]: 'abc'.replace('012345', '')
Out[3]: 'abc'
```

**Comment by mhvk** [comment]
I think I have a fix in #25515
